### ISSUE TITLE
Add --rm=dead or --rm=always to ease systemd integration

### ISSUE THREAD

**Issue by virtuald** [docker/docker#7245]
AFAIK, running a docker container 'properly' from systemd is very tricky, because of things like --rm and start/run/environment and other things. Various people have complained about this a bit, and various solutions exist -- all of which involve running a large shell script in your systemd unit file. I think I would really want the preferred way to run a container from systemd to be simple and not require shell scripting.

What about adding additional options to run --rm?

* `--rm=true/false`: same as before
* `--rm=dead`: if the name doesn't exist, do same as `--rm=true`. If a container with the name exists, removes the prior container if it was dead/not running. Raises an error if the current container is already running. Would make https://github.com/docker/docker/pull/6786 largely unnecessary
* `--rm=always`: runs the equivalent of `docker kill C; docker rm C; docker run --rm C`. See https://github.com/docker/docker/issues/4212. Honestly, I don't know if this should be added, you could get yourself into trouble with this.

The reason something like `--rm=dead` is required is because I cannot always guarantee that my container will be killed on reboot, such as during a power failure. Lacking an --rm=dead functionality, I have to write a shell script to do things like stop/rm/etc before I start my container. 

The 'proper' way of integrating throwaway containers with systemd would be greatly simplified, and look something like this:

    [Service]
    ExecStart=docker run --rm=dead --name=C container/name
    ExecStop=docker stop C

**Comment by cpuguy83** [comment]
The way I resolve this is by having an `ExecStartPre` that does the `docker rm -f` and ignores errors.

**Comment by cjbottaro** [comment]
Why did all work on this stall out?  I followed all the links and there is a lot of discussion, a lot of PRs, etc... and it all ends with comments like "So why was this closed?"

I also share the massive frustration of `docker run --rm` combined with Supervisor and Ansible.

**Comment by thaJeztah** [comment]
@cjbottaro have a look at https://github.com/docker/docker/pull/20848, which moves the `--rm` functionality from the client to the daemon

**Comment by cjbottaro** [comment]
Ahh, thank you.  Sometimes it's hard to navigate all the issues/pr links.

Here's my workaround in case anyone is interested in the meantime.  It's a `docker-run` shim that cleans up stopped containers first:

```python
#!/usr/bin/env python

import os
import sys
import subprocess

try:
    i = sys.argv.index("--name")
    name = sys.argv[i+1]
except ValueError:
    name == None

if "--rm" in sys.argv and name:
    with open(os.devnull, "w") as devnull:
        subprocess.call(["docker", "rm", name], stdout=devnull, stderr=devnull)

# sys.argv includes the program name (this script), so we copy and replace it.
# Note that the args must include "docker" as the first arg, despite it
# being redundant in the execvp call.
args = ["docker", "run"] + sys.argv[1:]

# docker run ...
os.execvp("docker", args)
```

Note that it only works when you use `--name some_name`.  It doesn't work if you do your name argument like `--name=some_name`

**Comment by icecrime** [comment]
I'm tentatively closing as it _should_ be solved by #20848. Please confirm and ping us to reopen if that's not the case!
